ambry-router/src/main/java/com.github.ambry.router/GetOperation.java

```diff
@@ -0,0 +1,204 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+import com.github.ambry.clustermap.ClusterMap;
+import com.github.ambry.clustermap.ReplicaId;
+import com.github.ambry.commons.BlobId;
+import com.github.ambry.commons.ResponseHandler;
+import com.github.ambry.commons.ServerErrorCode;
+import com.github.ambry.config.RouterConfig;
+import com.github.ambry.messageformat.MessageFormatFlags;
+import com.github.ambry.network.ResponseInfo;
+import com.github.ambry.protocol.GetOptions;
+import com.github.ambry.protocol.GetRequest;
+import com.github.ambry.protocol.PartitionRequestInfo;
+import com.github.ambry.utils.Time;
+import java.util.ArrayList;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An abstract class for a get operation.
+ * @param <T> the type of the result of this operation.
+ */
+abstract class GetOperation<T> {
+  final RouterConfig routerConfig;
+  final ClusterMap clusterMap;
+  final ResponseHandler responseHandler;
+  final FutureResult<T> operationFuture;
+  final Callback<T> operationCallback;
+  final BlobId blobId;
+  final Time time;
+  boolean operationCompleted = false;
+  Exception operationException;
+
+  static final Logger logger = LoggerFactory.getLogger(GetOperation.class);
+
+  /**
+   * Construct a GetOperation
+   * @param routerConfig the {@link RouterConfig} containing the configs for put operations.
+   * @param clusterMap the {@link ClusterMap} of the cluster
+   * @param responseHandler the {@link ResponseHandler} responsible for failure detection.
+   * @param blobIdStr the blobId of the associated blob in string form.
+   * @param futureResult the future that will contain the result of the operation.
+   * @param callback the callback that is to be called when the operation completes.
+   * @param time the {@link Time} instance to use.
+   * @throws RouterException if there is an error with any of the parameters, such as an invalid blob id.
+   */
+  GetOperation(RouterConfig routerConfig, ClusterMap clusterMap, ResponseHandler responseHandler, String blobIdStr,
+      FutureResult<T> futureResult, Callback<T> callback, Time time)
+      throws RouterException {
+    this.routerConfig = routerConfig;
+    this.clusterMap = clusterMap;
+    this.responseHandler = responseHandler;
+    this.operationFuture = futureResult;
+    this.operationCallback = callback;
+    this.time = time;
+    blobId = RouterUtils.getBlobIdFromString(blobIdStr, clusterMap);
+  }
+
+  /**
+   * Return the {@link FutureResult} associated with this operation.
+   * @return the {@link FutureResult} associated with this operation.
+   */
+  FutureResult<T> getFuture() {
+    return operationFuture;
```

Comment:
+1. one more minor comment: I have usually seen getters towards the end of the class in our code base. Any particular reason to have it in the beginning.